src/particle.cpp

```diff
@@ -0,0 +1,681 @@
+
+      if (!model::active_meshsurf_tallies.empty())
+        score_surface_tally(this, model::active_meshsurf_tallies);
+
+      // Clear surface component
+      surface_ = 0;
+
+      if (settings::run_CE) {
+        collision(this);
+      } else {
+        collision_mg(this);
+      }
+
+      // Score collision estimator tallies -- this is done after a collision
+      // has occurred rather than before because we need information on the
+      // outgoing energy for any tallies with an outgoing energy filter
+      if (!model::active_collision_tallies.empty()) score_collision_tally(this);
+      if (!model::active_analog_tallies.empty()) {
+        if (settings::run_CE) {
+          score_analog_tally_ce(this);
+        } else {
+          score_analog_tally_mg(this);
+        }
+      }
+
+      // Reset banked weight during collision
+      n_bank_ = 0;
+      n_bank_second_ = 0;
+      wgt_bank_ = 0.0;
+      for (int& v : n_delayed_bank_) v = 0;
+
+      // Reset fission logical
+      fission_ = false;
+
+      // Save coordinates for tallying purposes
+      r_last_current_ = this->r();
+
+      // Set last material to none since cross sections will need to be
+      // re-evaluated
+      material_last_ = C_NONE;
+
+      // Set all directions to base level -- right now, after a collision, only
+      // the base level directions are changed
+      for (int j = 0; j < n_coord_ - 1; ++j) {
+        if (coord_[j + 1].rotated) {
+          // If next level is rotated, apply rotation matrix
+          const auto& m {model::cells[coord_[j].cell]->rotation_};
+          const auto& u {coord_[j].u};
+          coord_[j + 1].u = u.rotate(m);
+        } else {
+          // Otherwise, copy this level's direction
+          coord_[j+1].u = coord_[j].u;
+        }
+      }
+
+      // Score flux derivative accumulators for differential tallies.
+      if (!model::active_tallies.empty()) score_collision_derivative(this);
+    }
+
+    // If particle has too many events, display warning and kill it
+    ++n_event;
+    if (n_event == MAX_EVENTS) {
+      warning("Particle " + std::to_string(id_) +
+        " underwent maximum number of events.");
+      alive_ = false;
+    }
+
+    // Check for secondary particles if this particle is dead
+    if (!alive_) {
+      // If no secondary particles, break out of event loop
+      if (simulation::secondary_bank.empty()) break;
+
+      this->from_source(&simulation::secondary_bank.back());
+      simulation::secondary_bank.pop_back();
+      n_event = 0;
+
+      // Enter new particle in particle track file
+      if (write_track_) add_particle_track();
+    }
+  }
+
+  #ifdef DAGMC
+  if (settings::dagmc) simulation::history.reset();
+  #endif
+
+  // Finish particle track output.
+  if (write_track_) {
+    write_particle_track(*this);
+    finalize_particle_track(*this);
+  }
+}
+
+void
+Particle::cross_surface()
+{
+  int i_surface = std::abs(surface_);
+  // TODO: off-by-one
```

Comment:
It it doesn't already exist, I would consider writing a test that fails because of this TODO item, so whoever writes it knows they did it right.